fetch/api/basic/multi-globals.html

```diff
@@ -0,0 +1,111 @@
+  assert_true(isObjectFromGlobal(response, global), `response should use the ${globalInWords} global`);
+  assert_true(isObjectFromGlobal(response.headers, global), `response.headers should use the ${globalInWords} global`);
+  if (response.body !== null) {
+    assert_true(isObjectFromGlobal(response.body, global), `response.body should use the ${globalInWords} global`);
+  }
+}
+
+window.addEventListener("load", () => {
+  test(t => {
+    // This will invoke the method from currentFrame, but with incumbentFrame as the incumbent, and
+    // relevantFrame as the relevant.
+    const errorResponse = incumbentFrame.contentWindow.runResponseError();
+    assert_equals(errorResponse.status, 0);
+    assert_equals(errorResponse.body, null);
+
+    assert_response(errorResponse, currentFrame.contentWindow, "current");
+  }, "Response.error()");
+
+  test(t => {
+    const redirectResponse = incumbentFrame.contentWindow.runResponseRedirect("/", 302);
+    assert_equals(redirectResponse.status, 302);
+    assert_equals(redirectResponse.body, null);
+
+    assert_response(redirectResponse, currentFrame.contentWindow, "current");
+  }, "Response.redirect()");
+
+  promise_test(async t => {
+    const responsePromise = incumbentFrame.contentWindow.runFetch("data:,x");
+
+    assert_true(isObjectFromGlobal(responsePromise, relevantFrame.contentWindow), "fetch()'s promise should use the relevant global");
+
+    const response = await responsePromise;
+    assert_not_equals(response.body, null);
+
+    assert_response(response, relevantFrame.contentWindow, "relevant");
+  }, "fetch()");
+
+  bodyMethods.forEach(method => {
+    promise_test(async t => {
+      // This data URL was chosen as it works for formData() and json()
+      const response = await incumbentFrame.contentWindow.runFetch("data:application/x-www-form-urlencoded,[]");
+
+      const bodyPromise = currentFrame.contentWindow.Response.prototype[method].call(response);
+
+      assert_true(isObjectFromGlobal(bodyPromise, relevantFrame.contentWindow), `Response's ${method}()'s promise should use the relevant global`);
+
+      // Primitive types don't have a global
+      if (method === "text") {
+        return;
+      }
+
+      const body = await bodyPromise;
+
+      assert_true(isObjectFromGlobal(body, relevantFrame.contentWindow), `Response's ${method}()'s object should use the relevant global`);
+    }, `Response's ${method}()`);
+  });
+
+  promise_test(async t => {
+    const response = await incumbentFrame.contentWindow.runFetch("data:,");
+
+    // Consume the response
+    await response.arrayBuffer();
+
+    // Verify rejected promises
+    return Promise.all(bodyMethods.map(method => {
+      const rejectedPromise = currentFrame.contentWindow.Response.prototype[method].call(response);
+      assert_true(isObjectFromGlobal(rejectedPromise, currentFrame.contentWindow), `Response's ${method}()'s rejected promise should use the current global`)
```

Comment:
Shouldn't this be relevant? The exception should be current, but the promise should be relevant, I believe.

I'm not too happy with that inconsistency myself, but the problem is overconstrained:

- We decided the web platform uses relevant by default
- ES uses current by default
- We decided Web IDL exception creation and throwing matches ES
- Promises are created explicitly and we decided they should match the web platform

Reply:
Yeah, I wanted to discuss this. I think the problem is that browsers let the IDL layer construct the promise. We discussed this previously about the difference between specifications using "throw" and "return a rejected promise" I think. As far as I can tell browsers might use "throw". Or am I missing something?

cc @TimothyGu @yuki3

Reply:
Yeah, in Chrome at least a lot of code uses throw and the bindings layer converts it into a rejected promise. (I'm not sure how this works for in-parallel cases; maybe it's just for synchronously-detectable error conditions?) But I think the bindings layer could still convert it into a rejected promise in the relevant realm wrapping an exception in the current realm.

Reply:
I see, so you want browsers to change their behavior? I suspect that would also require changes to IDL.

Reply:
I mean, what I really want is to use current everywhere, but I gave up on that due to bz's pushback.

My fallback position is to maintain as much local consistency as possible, and as part of that I think it's important that rejected promises and fulfilled promises have the same realm. We could make that realm either current or relevant. Relevant seems to maximize consistency and minimize confusion in giving spec-authoring guidance more. If we did current we'd have to update our guidance to something like "use relevant, except for promises; use current for those". That seems a bit bad?

I guess we would need IDL changes; https://heycam.github.io/webidl/#dfn-create-operation-function implicitly uses the current Realm (via `%Promise%` syntax).

Reply:
About Chrome's bindings, nothing to add to domenic's understanding.  In synchronous cases, "throw an exception -> convert it to a rejected promise" is done in bindings layer.  Generally bindings layer doesn't take care of "in parallel" cases, but provides some utility functions.

As long as it's consistent and clean, spec authors won't be confused, and no trouble at the end users, I'm fine to change the realm of (rejected) promises to the current realm.  I'm fine with a proposal of: interface => relevant, others => current.

Just my two cents.

Reply:
Thanks @yuki3. I guess we need to solve https://github.com/heycam/webidl/issues/135 first then.